x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/ClusterDeprecationChecks.java

```diff
@@ -91,6 +96,54 @@ static DeprecationIssue checkTemplatesWithTooManyFields(ClusterState state) {
         return null;
     }
 
+    /**
+     * Check templates that use `_field_names` explicitly, which was deprecated in https://github.com/elastic/elasticsearch/pull/42854
+     * and will throw an error on new indices in 8.0
+     */
+    static DeprecationIssue checkTemplatesWithFieldNamesDisabled(ClusterState state) {
+        Set<String> templatesContainingFieldNamed = new HashSet<>();
+        state.getMetaData().getTemplates().forEach((templateCursor) -> {
+            String templateName = templateCursor.key;
+            templateCursor.value.getMappings().forEach((mappingCursor) -> {
+                Map<String, Object> map = XContentHelper.convertToMap(mappingCursor.value.compressedReference(), false, XContentType.JSON)
+                        .v2();
+                if (mapContainsFieldNamesDisabled(map)) {
+                    templatesContainingFieldNamed.add(templateName);
+                }
+            });
+        });
+
+        if (templatesContainingFieldNamed.isEmpty() == false) {
+            return new DeprecationIssue(DeprecationIssue.Level.CRITICAL, "Index templates contain _field_names settings.",
+                    "https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-8.0.html#fieldnames-enabling",
+                    "Index templates " + templatesContainingFieldNamed + " use the deprecated `enable` setting for the `"
+                            + FieldNamesFieldMapper.NAME + "` field. Using this setting in new index mappings will throw an error "
+                                    + "in the next major version and needs to be removed from existing mappings and templates.");
+        }
+        return null;
+    }
+
+    /**
+     * check for "_field_names" entries in the map that contain another property "enabled" in the sub-map
+     */
+    static boolean mapContainsFieldNamesDisabled(Map<?, ?> map) {
```

Comment:
Do we need to use a recursive function here, or could we just check for it directly since we know `_field_names` will be at the top level of the mapping? Also, there is a convenient method `IndexDeprecationChecks#findInPropertiesRecursively` for iterating over a mapping.

Reply:
I opted for the recursive function just to make sure we catch all type-vs-no-type cases (see also the test cases). If we always have (or don't have) a type-level in the internally stored mappings in the templates, then we can probably do without but I wanted to check with you on this, it wasn't completely clear to me from reading the code. The mapping maps are quite forgiving for what you store in them, so it depends on the caller and I didn't completely follow the internal logic.

> there is a convenient method `IndexDeprecationChecks#findInPropertiesRecursively`

I saw and tried to use this, but ran into problems because it assumes things are stored under a `properties` key which I think the `"_field_nams" : { "enabled" : ...}` is not (correct me if I'm wrong).

Let me if you know for sure how the internal map structure is supposed to look like (at which level to expect the "_field_names" key) and I can update the tests accordingly and try making this non-recursive.

Reply:
> I saw and tried to use this, but ran into problems because it assumes things are stored under a properties key 

Ah, I missed this -- then `findInPropertiesRecursively` is indeed not a good fit. 

> Let me if you know for sure how the internal map structure is supposed to look like

I left some comments on your individual questions, let me know if anything is still unclear!